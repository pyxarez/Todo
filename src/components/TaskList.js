import React from 'react';

import TextBox from './TextBox';
import './TaskList.res/style.css';

TaskList.propTypes = {
  children: React.PropTypes.oneOfType(
    [
      React.PropTypes.arrayOf(React.PropTypes.element.isRequired),
      React.PropTypes.bool,
      React.PropTypes.object
    ]
  ),
  onClick: React.PropTypes.func.isRequired
}

export default function TaskList({ onClick, children }) {
  return (
    <div className="my-tasklist-component">
      <TextBox
        placeholder="Enter task title"
        onClick={onClick}/>
      <div className="tasklist">
        {children}
      </div>
    </div>
  );
}


