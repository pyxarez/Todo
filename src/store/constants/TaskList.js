export const ADD_TASK = 'ADD_TASK';
export const CHECK_DONE = 'CHECK_DONE';
export const GET_TASKS = 'GET_TASKS';
export const SAVE_TASK_CHANGES = 'SAVE_TASK_CHANGES';
