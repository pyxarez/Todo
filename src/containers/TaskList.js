import React, { Component, PropTypes } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { addTask, getTasks } from '../store/actions/TaskListActions';
import { getProgress } from '../store/actions/ProgressActions';

import TaskList from '../components/TaskList';
import TaskContainer from './Task';

export class TaskListContainer extends Component {
  static propTypes = {
    showDone: PropTypes.bool.isRequired,
    tasks: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.node.isRequired,
      title: PropTypes.string.isRequired,
      isDone: PropTypes.bool.isRequired,
      description: React.PropTypes.string.isRequired
    })),
    addTask: PropTypes.func.isRequired,
    getTasks: PropTypes.func.isRequired,
    getProgress: React.PropTypes.func.isRequired,
    URLParams: PropTypes.shape({
      id: PropTypes.number,
      category: PropTypes.string,
      filter: PropTypes.string
    })
  }

  handleAddNewTaskClick = (title) => {
    const {
      addTask,
      getProgress
    } = this.props;
    const id = this.props.URLParams.id;

    return addTask(id, title)
      .then(() => { getProgress();})
      .catch(e => { alert(e); });
  }

  componentDidMount = () => {
    const { id } = this.props.URLParams;
    
    if (typeof id !== 'undefined') {
      this.props.getTasks(id);
    }
  }

  componentWillReceiveProps = (nextProps) => {
    const nextId = nextProps.URLParams.id;
    const prevId = this.props.URLParams.id

    if ( (typeof nextId !== 'undefined') && (nextId !== prevId) ){
      this.props.getTasks(nextId);
    }
  }

  _createTasks = () => {
    const {
      showDone,
      tasks
    } = this.props;
    const {
      id: categoryId,
      filter: keyWord = ''
    } = this.props.URLParams;

    const sortedTaskList = showDone
      ? tasks
      : tasks.filter(task => !task.isDone);

    return sortedTaskList
      .filter((task) => task.title
        .toLowerCase()
        .includes(keyWord
          .toLowerCase()
          .trim()
         )
       )
      .map((task, index) => (
          <TaskContainer
            key={task.id}
            categoryId={categoryId}
            task={task}/>
        )
      );
  }

  render() {
    const { tasks } = this.props;
    return (
      <TaskList onClick={this.handleAddNewTaskClick}>
        {tasks.length > 0 && this._createTasks()}
      </TaskList>
    );
  }
}

const mapStateToProps = ({ tasks }) => {
  return {
    tasks
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addTask: bindActionCreators(addTask, dispatch),
    getTasks: bindActionCreators(getTasks, dispatch),
    getProgress: bindActionCreators(getProgress, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(TaskListContainer);