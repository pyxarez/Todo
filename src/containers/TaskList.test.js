import React from 'react';
import { shallow, mount } from 'enzyme';

import { TaskListContainer } from './TaskList';
import TaskList from '../components/TaskList';

const setup = () => {
  const props = {
    showDone: true,
    tasks: [
      {
        id: 0,
        title: 'New Title',
        isDone: false,
        description: '1 2 34 5 6 7'
      },
      {
        id: 1,
        title: 'New second Title',
        isDone: true,
        description: '1 2 34 5 6 7'
      },
      {
        id: 3,
        title: 'New Title yeaaa',
        isDone: false,
        description: '1 2 34 5 6 7'
      },
    ],
    addTask: jest.fn((id, title) => {
      return new Promise((resolve, reject) => {
          if ((id === undefined) || !title) reject('Error');
          resolve('I\'m born!');
        });
      }
    ),
    getProgress: jest.fn(),
    getTasks: jest.fn(),
    URLParams: {
      id: 0,
      category: 'Some category',
      filter: 'Task filter'
    }
  }
  const wrapper = shallow(<TaskListContainer {...props}/>);

  return {
    wrapper,
    props
  }
};

describe('Containers :: TaskListContainer', () => {
  it('should render self and subcomponents', () => {
    const { wrapper } = setup();
    
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component without done tasks', () => {
    const { props } = setup();
    const newProps = {
      ...props,
      showDone: false
    };
    
    const wrapper = shallow(<TaskListContainer {...newProps}></TaskListContainer>)
    expect(wrapper).toMatchSnapshot();
  });

  it('should call addTask', () => {
    const { props, wrapper } = setup();
    const newTitle = 'New title';

    wrapper.find(TaskList).prop('onClick')(newTitle);
    expect(props.addTask)
      .toBeCalledWith(props.URLParams.id, newTitle);
  });

  it('should call getProgress', () => {
    const { props, wrapper } = setup();
    const newTitle = 'New title';
    
    return expect(wrapper.find(TaskList).prop('onClick')(newTitle)
      .then(() => {
        expect(props.getProgress).toBeCalled();
      })).resolves;
  });

  it('should call alert after failure in addTask function', () => {
    const { props } = setup();
    const newProps = {
      ...props,
      URLParams: {
        id: undefined,
      } 
    };
    const wrapper = shallow(<TaskListContainer {...newProps}/>);
    window.alert = jest.fn();
  
    
    return expect(wrapper.find(TaskList).prop('onClick')()
      .then(() => {
        expect(window.alert).toBeCalled();
      })).rejects;
  });

  it('should call getTasks after receiving new props', () => {
    const { wrapper, props } = setup();
    const newProps = {
      URLParams: {
        id: 1
      }
    };

    wrapper.instance().componentWillReceiveProps(newProps);

    expect(props.getTasks)
      .toBeCalledWith(newProps.URLParams.id);
  });

  it('should not call getTasks after receiving new props', () => {
    const { wrapper, props } = setup();
    const nextProps = {
      URLParams: {
        id: 0
      }
    };

    wrapper.instance().componentWillReceiveProps(nextProps);

    expect(props.getTasks)
      .not
      .toBeCalled();
  });
});