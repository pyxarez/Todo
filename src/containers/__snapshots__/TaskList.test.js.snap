exports[`Containers :: TaskListContainer should render component without done tasks 1`] = `
<TaskList
  onClick={[Function]} />
`;

exports[`Containers :: TaskListContainer should render self and subcomponents 1`] = `
<TaskList
  onClick={[Function]} />
`;
